### DISCUSSION TITLE
Can't register a singleton with a SoapClient in the constructor

### DISCUSSION THREAD

**Discussion by MarianNapi** [laravel/framework#33146] [General]
- Laravel Version: 7.14.1
- PHP Version: 7.3.18

I've created a new ServiceProvider and added it to the app.php. I then registered a singleton with a test class which has a SoapClient in it's constructor:

`        $this->app->singleton(
            Test::class,
            function () {
                return new Test(new SoapClient(null));
            }
        );`

I then get this error when trying to get the class like this:

`app(Test::class)`

> Illuminate\Contracts\Container\BindingResolutionException : Unresolvable dependency resolving [Parameter #0 [ <required> $wsdl ]] in class SoapClient

It doesn't matter wether I use null or a real wsdl as SoapClient parameter, the error stays the same.

**Comment by driesvints** [comment]
Moving to discussions

**Comment by mfn** [comment]
Because the `SoapClient` constructor requires the WSDL, see https://www.php.net/manual/en/soapclient.construct.php

**Reply by t-heuser** [reply]
No, this is not true. The parameter can also be null, as seen [here](https://www.php.net/manual/en/soapclient.soapclient.php).
But as @MarianNapi wrote: 

> It doesn't matter wether I use null or a real wsdl as SoapClient parameter, the error stays the same.

So he even tried it with a real WSDL, but he still gets the same error.

Does the problem perhaps arise from the fact that the paramter $wsdl is of the type mixed and laravel cant resolve that correctly?

**Reply by mfn** [reply]
Good catch!

Then the only other reason I can think of is that the registration is not properly picked up, because it would only try auto-wiring if it can't find an explicit bind, which however was given.

- is `Test::class` referring to the same class, i.e. is the import correct?
- Is it a deferred provider and maybe `provides` isn't set up correctly?

**Reply by MarianNapi** [reply]
When using the GuzzleClient in the Test class and then registering the class with
 `$this->app->singleton( Test::class, function () { return new Test(new Client()); } );` 
it's working completely fine - so I don't think the import is the problem. 

The provider doesn't implement the DeferrableProvider so as far as I understood the docs the `provides` method isn't needed here
